src/hawkbit-client.c

```diff
@@ -369,67 +509,106 @@ static void json_build_status(JsonBuilder *builder, const gchar *id, const gchar
+
+        if (!g_strcmp0(finished, "failure"))
+                g_warning("%s", detail);
+        else
+                g_message("%s", detail);
+
+        builder = json_build_status(id, detail, finished, execution, NULL);
+
+        res = rest_request(POST, url, builder, NULL, error);
+        if (!res)
+                g_prefix_error(error, "Failed to report \"%s\" feedback: ", detail);
 
-        int status = rest_request(POST, url, builder, NULL, error);
-        g_debug("Feedback status: %d, URL: %s", status, url);
-        g_object_unref(builder);
-        return (status == 200);
+        return res;
 }
 
 /**
- * @brief Send progress feedback to hawkBit.
+ * @brief Send progress feedback to hawkBit (finished=none, execution=proceeding).
+ *
+ * @param[in]  url    hawkBit URL used for request
+ * @param[in]  id     hawkBit action ID or NULL (configData usecase)
+ * @param[in]  detail Detail message or NULL (configData usecase)
+ * @param[out] error  Error
+ * @return TRUE if feedback was sent successfully, FALSE otherwise (error set)
  */
-static gboolean feedback_progress(const gchar *url, const gchar *id, const gchar *detail, GError **error)
+static gboolean feedback_progress(const gchar *url, const gchar *id, const gchar *detail,
+                                  GError **error)
 {
-        JsonBuilder *builder = json_builder_new();
-        json_build_status(builder, id, detail, "none", "proceeding", NULL);
+        gboolean res = FALSE;
+
+        g_return_val_if_fail(url, FALSE);
+        g_return_val_if_fail(id, FALSE);
+        g_return_val_if_fail(detail, FALSE);
+        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
 
-        int status = rest_request(POST, url, builder, NULL, error);
-        g_debug("Feedback progress status: %d, URL: %s", status, url);
-        g_object_unref(builder);
-        return (status == 200);
+        res = feedback(url, id, detail, "none", "proceeding", error);
+        g_prefix_error(error, "Progress feedback: ");
+
+        return res;
 }
 
 /**
  * @brief Get polling sleep time from hawkBit JSON response.
+ *
+ * @param[in] root JsonNode* with hawkBit response
+ * @return time to sleep in seconds, either from JSON or (if not found) from config's retry_wait
  */
 static long json_get_sleeptime(JsonNode *root)
 {
-        gchar *sleeptime_str = NULL;
+        g_autofree gchar *sleeptime_str = NULL;
+        g_autoptr(GError) error = NULL;
         struct tm time;
-        long poll_sleep_time;
 
-        sleeptime_str = json_get_string(root, "$.config.polling.sleep", NULL);
+        g_return_val_if_fail(root, hawkbit_config->retry_wait);
+
+        sleeptime_str = json_get_string(root, "$.config.polling.sleep", &error);
         if (!sleeptime_str) {
-                poll_sleep_time = hawkbit_config->retry_wait;
-                goto out;
+                g_warning("Polling sleep time not found: %s", error->message);
```

Comment:
I like that, will use it :)

Reply:
I answered with
> Both the JsonNode and the sleep time are always expected. So if these are not there, that's an error. I'll return 0 for both cases.

Now I'm unsure about this. Should we handle such a case (which should never happen according to the DDI API documentation) with a warning and carry on with `hawkbit_config->retry_wait`?

Reply:
> I answered with

Ok, was lost in review comments.

I'd say it is okay to abort if the API breaks unexpectedly, but then I would semantically not use `g_return_val_if_fail()` but `g_error()` or similar instead.

Reply:
I could add a `GError**` argument to `json_get_sleeptime()`. Looking at the function calling `json_get_sleeptime()` (being `hawkbit_pull_cb()`), this would probably lead to an immediate..
```
g_warning("%s", error->message);
g_clear_error(&error);
```
Then we would wait for either the last sleep time we got or `hawkbit_config->retry_wait`.

I don't know if that's worth the hassle. Maybe we should go back to the initial way and simply emit a warning here directly and live with `hawkbit_config->retry_wait`?

Reply:
> I don't know if that's worth the hassle. Maybe we should go back to the initial way and simply emit a warning here directly and live with hawkbit_config->retry_wait?

I'd be fine with that too, yes.

Reply:
I've simply re-applied the (slightly modified) warning variant now to resolve this PR.